### PR TITLE
Some structure in DAChain for randomized subchain lenghts

### PR DESCRIPTION
**Pull request by louisekluge**
Work in progress: implementing randomized subchain lengths and data structures that later on allow to calculate the (asymptoticaly unbiased) multilevel estimator.

### CODE REVIEW

**Comment by louisekluge** [comment]
With the added structure I am able to run DAChain from the chain level, with randomized subchain lengths. However, there seem to be some conflicts with sampler.py, that I don't fully understand yet. Also, in initial tests with randomized subchain lengths, the acceptance rates displayed for the fine chain swiftly go to zero.  Might that have something to do with how the acceptance rates are computed? This happens also, when I increase the subsampling rate substantially, whereas increasing the subsampling rate without RSL is largely invariant with increased subsampling rate.

File: tinyDA/chain.py [hunk 1]
```diff
@@ -149,6 +149,9 @@ class DAChain:
         Transition kernel for coarse MCMC proposals.
     subsampling_rate : int
         The subsampling rate for the coarse chain.
+    randomized_subchain_length : bool, optional
```
Comment (mikkelbue) [inline tinyDA/chain.py:152]: Can we rename this variable to `randomize_subchain_length`, so in active form, without the "d"?

File: tinyDA/link.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ class Link:
         The model output
     likelihood : float
         The log-likelihood of the data, given the parameters.
-    qoi
+    qoi : numpy.ndarray
```
Comment (mikkelbue) [inline tinyDA/link.py:17]: `qoi` can be of `any` type.

**Review by mikkelbue — COMMENTED** [review]
@louisekluge If this isn't working, it's probably because the acceptance ratios haven't been adapted. I think you would have to change the method `_get_state_dependent_acceptance`, and also the lines 351-364:
```
                if np.random.random() < alpha_2:
                    self.chain_fine.append(proposal_link_fine)
                    self.accepted_fine.append(True)
                    self.chain_coarse.append(self.chain_coarse[-1])
                    self.accepted_coarse.append(True)
                    self.is_coarse.append(False)
                else:
                    self.chain_fine.append(self.chain_fine[-1])
                    self.accepted_fine.append(False)
                    self.chain_coarse.append(
                        self.chain_coarse[-(self.subsampling_rate + 1)]
                    )
                    self.accepted_coarse.append(False)
                    self.is_coarse.append(False)
```
Currently, it is creating a fine link from a random coarse, which is what we want, but it is still assuming that the coarse state is the last link in the coarse chain when computing the acceptance rate and adding the link to the chain.

File: tinyDA/link.py (hunk 2 above)
Comment (louisekluge) [reply tinyDA/link.py:17]: Ah, I see. I didn't get that

Comment (louisekluge) [reply tinyDA/link.py]: I set it to any in the recent commit, hope that is what you meant

**Comment by louisekluge** [comment]
I tried implementing all your suggestions from the last comment and it's looking good so far. Acceptance rates on the fine chain look good, even a little high but similar to the runs without randomized subchain lengths.

**Comment by mikkelbue** [comment]
Nice! I think we have to think a little about the implementation, since there is a huge `if-else` block now, which is a little hard to read. One option would to have a `promoted_coarse` list going even when we don't use randomized subchain lengths. Then have the acceptance tests address the latest link in that list no matter what, and wouldn't have to have separate logic for each case. That way we could avoid quite a few places in the existing code as well, where we use the index `self.chain_coarse[-(self.subsampling_rate + 1)]`, which would make it cleaner.

**Comment by mikkelbue** [comment]
> That way we could avoid quite a few places in the existing code as well, where we use the index self.chain_coarse[-(self.subsampling_rate + 1)], which would make it cleaner.

@louisekluge Thinking about this a little more, I don't think what I said makes sens, because `self.chain_coarse[-(self.subsampling_rate + 1)]` refers to the state not the proposal.

Nevertheless, I have some ideas to make this a little cleaner:

1. Let's use the `self.promoted_coarse` to hold the proposal for the coarse, whether we use `randomize_subchain_length` or not. Then we can get rid of that entire `if-else` construction that was introduced in your previous commit. If we always have the proposal in as the latest item in that list, we can always do `self.chain_coarse.append(self.promoted_coarse[-1])`.
2. This means we will always have to add something to the `self.promoted_coarse` at the end of coarse sampling. When `randomize_subchain_length=False`, this would always be the latest coarse sample. This could be done inside `_sample_coarse`, to make `sample` a little cleaner.
3. Instead of having an if-statement to check `randomize_subchain_length`  in **every** iteration of the `_sample_coarse`, we could use this design pattern for the `__init__` function:
```
if self.randomize_subchain_length:
    # this private method returns np.random.randint(-self.subsampling_rate,0)
    self._get_proposal_index = self._get_random_proposal_index()
else:
    # this private method always returns -1
    self._get_proposal_index = self._get_fixed_proposal_index()
```
We would then simply trigger `self._get_proposal_index` at the end of `_sample_coarse`:
```
proposal_index =  self._get_proposal_index()
self.promoted_coarse.append(self.chain_coarse[proposal_index])
```
4. Minor point `if self.randomize_subchain_length==True:` (and similar) can be simplified to `if self.randomize_subchain_length:`. Similarly `if self.randomize_subchain_length==False:` should be ``if not self.randomize_subchain_length:`

Let me know if you have any questions about this, and we can also discuss it when we meet next time.

**Comment by louisekluge** [comment]
I think there's a mistake (on my side) in _get_state_independent_acceptance, in the sense that we should not look at the last element of the last coarse chain but at the prior element of promoted_coarse. 
`    def _get_state_independent_acceptance(self, proposal_link_fine):
        # compute the state-independent delayed acceptance probability.
        if self.randomize_subchain_length==True:
            alpha_2 = np.exp(
                proposal_link_fine.posterior
                - self.chain_fine[-1].posterior
                **+ self.chain_coarse[-(self.subsampling_rate + 1)].posterior**
                - self.promoted_coarse[-1].posterior
            )
        else:
            alpha_2 = np.exp(
                proposal_link_fine.posterior
                - self.chain_fine[-1].posterior
                + self.chain_coarse[-(self.subsampling_rate + 1)].posterior
                - self.chain_coarse[-1].posterior
            )
        return alpha_2`


Is the formulation below correct?

`    def _get_state_independent_acceptance(self, proposal_link_fine):
        # compute the state-independent delayed acceptance probability.
          alpha_2 = np.exp(
              proposal_link_fine.posterior
              - self.chain_fine[-1].posterior
              + self.promoted_coarse[-2].posterior
              - self.promoted_coarse[-1].posterior
          )
return alpha_2
`

**Comment by louisekluge** [comment]
I assume that randomize_subchain_length is now functional, it passes my tests so far. I will now look at the example and the estimator.

File: tinyDA/chain.py [hunk 3]
```diff
@@ -347,18 +384,18 @@ def sample(self, iterations, progressbar=True):
...
-                    self.chain_coarse.append(self.chain_coarse[-1])
+                    self.chain_coarse.append(self.promoted_coarse[-1])
                     self.accepted_coarse.append(True)
                     self.is_coarse.append(False)
                 else:
                     self.chain_fine.append(self.chain_fine[-1])
                     self.accepted_fine.append(False)
                     self.chain_coarse.append(
-                        self.chain_coarse[-(self.subchain_length + 1)]
+                        self.chain_coarse[-(self.subchain_length+1)]
```
Comment (mikkelbue) [inline tinyDA/chain.py:398]: It looks like you have removed some spaces here that makes it look like there is a change but there isn't,

File: tinyDA/chain.py [hunk 4]
```diff
@@ -217,12 +229,15 @@ def __init__(
...
         self.subchain_length = subchain_length
+        self.randomize_subchain_length = randomize_subchain_length 
 
         # set up lists to hold coarse and fine links, as well as acceptance
         # accounting
         self.chain_coarse = []
         self.accepted_coarse = []
         self.is_coarse = []
+        self.promoted_coarse = []
+        self.random_subchain_length = []
```
Comment (mikkelbue) [inline tinyDA/chain.py:240]: It's a semantic point but, since we are using this list to keep the lengths, whether they are random or not, maybe we can just call this list `self.subchain_lengths`? With an "s" in the end to differentiate it from `self.subchain_length`.

File: tinyDA/chain.py [hunk 5]
```diff
@@ -336,9 +369,13 @@ def sample(self, iterations, progressbar=True):
...
                 # when subsampling is complete, create a new fine link from the
                 # previous coarse link.
+                proposal_index = self._get_proposal_index()
                 proposal_link_fine = self.posterior_fine.create_link(
-                    self.chain_coarse[-1].parameters
+                    self.chain_coarse[proposal_index].parameters
                 )
+                self.promoted_coarse.append(self.chain_coarse[proposal_index]) 
+                # add effective subchain lenght to list
+                self.random_subchain_length.append(proposal_index+self.subchain_length)
```
Comment (mikkelbue) [inline tinyDA/chain.py:378]: Please use `black` to format the code, this should be
```
self.random_subchain_length.append(proposal_index + self.subchain_length)
```

File: tinyDA/chain.py [hunk 6]
```diff
@@ -380,7 +417,14 @@ def _sample_coarse(self):
...
         for j in range(self.subchain_length):
             # draw a new proposal, given the previous parameters.
-            proposal = self.proposal.make_proposal(self.chain_coarse[-1])
+            if j==0:# use correct starting point for subchain
+                if self.accepted_fine[-1]:
+                    proposal = self.proposal.make_proposal(self.promoted_coarse[-1])
+                else: #if fine chain did not accept, jump back to start of last chain 
+                    proposal = self.proposal.make_proposal(self.chain_coarse[-(self.subchain_length+1)])
+
+            else:
```
Comment (mikkelbue) [inline tinyDA/chain.py:426]: I don't think this is necessary, since the latest coarse link should already be the correct one, according to the acceptance check on the finer level.

File: tinyDA/chain.py [hunk 7]
```diff
@@ -397,7 +441,7 @@ def _sample_coarse(self):
                 self.accepted_coarse.append(True)
                 self.is_coarse.append(True)
             else:
-                self.chain_coarse.append(self.chain_coarse[-1])
+                self.chain_coarse.append(self.promoted_coarse[-1]) # adapted to use promoted_coarse
```
Comment (mikkelbue) [inline tinyDA/chain.py:444]: I don't think this is correct. The coarse should always add its own previous link if the proposal is rejected.

File: tinyDA/link.py [hunk 8]
```diff
@@ -46,3 +46,12 @@ def __init__(self, parameters, prior, model_output, likelihood, qoi=None):
...
         self.posterior = self.prior + self.likelihood
+
+
+class RandomProposal:
+    """Separate object for each sample on the fine level, that points to the 
+    proposal that was promoted by the subchain.
+    
+    Attributes
+    ----------
+    """
```
Comment (mikkelbue) [inline tinyDA/link.py:57]: This doesn't seem to do anything.

File: tinyDA/chain.py [hunk 9]
```diff
@@ -533,6 +586,7 @@ def __init__(
         subchain_lengths,
         initial_parameters=None,
         adaptive_error_model=None,
+        #randomize_subchain_length=False,
```
Comment (mikkelbue) [inline tinyDA/chain.py:589]: This is commented out.

File: tinyDA/chain.py [hunk 10]
```diff
@@ -489,6 +533,15 @@ def _update_error_model(self):
             )
         self.chain_coarse[-1] = self.posterior_coarse.update_link(self.chain_coarse[-1])
 
+    def _get_random_proposal_index(self):
+        random_proposal_index = np.random.randint(-self.subchain_length,0) #prob does not correspond to correct state
+        return random_proposal_index
+    
+    def _get_fixed_proposal_index(self):
+        fixed_proposal_index = -1
+        return fixed_proposal_index
```
Comment (mikkelbue) [inline tinyDA/chain.py:542]:
```suggestion
        return -1
```

File: tinyDA/chain.py [hunk 11]
```diff
@@ -436,7 +481,6 @@ def _get_state_dependent_acceptance(self, proposal_link_fine):
                 self.chain_coarse[-1].posterior + q_y_x,
```
Comment (mikkelbue) [inline tinyDA/chain.py:481]: I think this needs to be `self.promoted_coarse[-1].posterior + q_y_x`

**Review by mikkelbue — CHANGES_REQUESTED** [review]
I have a couple of more comments, and some big-picture things:
* I don't think we need to make any changes to `_sample_coarse`, actually.
* There are still a few places, outside of `_sample_coarse` with reference to `self.chain_coarse[-1]` in place of `self.promoted_coarse[-1]`.

Please use the "Files changed" tab on the PR page on GitHub to look over your changes. Some of the lines that appear to have changes have only had their formatting changed, and I prefer if there are only actual changes highlighted.

**Comment by mikkelbue** [comment]
@louisekluge Please see my review above.

File: tinyDA/chain.py (hunk 9 above)
Comment (louisekluge) [reply tinyDA/chain.py:589]: Right, I have removed it as we are not thinking about MLDA in this pull request.

File: tinyDA/link.py (hunk 8 above)
Comment (louisekluge) [reply tinyDA/link.py:57]: Oh yes, this seems to be a remnant of an idea that I did not go through with. I removed it.

File: tinyDA/chain.py (hunk 7 above)
Comment (louisekluge) [reply tinyDA/chain.py:444]: So, it should be elf.chain_coarse.append(self.chain_coarse[-(self.subchain_length+1)]), right?

Comment (mikkelbue) [reply tinyDA/chain.py]: No, for running the coarse chain, if the sample is rejected, it should always just be `self.chain_coarse[-1]`.

If you would do `self.chain_coarse[-(self.subchain_length+1)]`, then if there had been an accepted sample before, but we now reject, then that previously accepted sample would be disregarded.

File: tinyDA/chain.py [hunk 12]
```diff
@@ -336,9 +366,13 @@ def sample(self, iterations, progressbar=True):
...
                 # when subsampling is complete, create a new fine link from the
                 # previous coarse link.
+                proposal_index = self._get_proposal_index()
                 proposal_link_fine = self.posterior_fine.create_link(
-                    self.chain_coarse[-1].parameters
+                    self.chain_coarse[proposal_index].parameters
                 )
+                self.promoted_coarse.append(self.chain_coarse[proposal_index]) 
+                # add effective subchain lenght to list
+                self.subchain_lengths.append(proposal_index + self.subchain_length)
```
Comment (mikkelbue) [inline tinyDA/chain.py:375]: I think the `subchain_length` here will actually be `proposal_index + self.subchain_length + 1`, since if for example we have `proposal_index == -self.subchain_length` (the shortest possible length), the specific `subchain_length` should still be 1. I think this is a somewhat semantic point, but it should also make constructing the estimator easier down the line.

File: tinyDA/chain.py [hunk 13]
```diff
@@ -397,7 +431,7 @@ def _sample_coarse(self):
                 self.accepted_coarse.append(True)
                 self.is_coarse.append(True)
             else:
-                self.chain_coarse.append(self.chain_coarse[-1])
+                self.chain_coarse.append(self.chain_coarse[-(self.subchain_length+1)]) # adapted to use promoted_coarse
```
Comment (mikkelbue) [inline tinyDA/chain.py:434]:
```suggestion
                self.chain_coarse.append(self.chain_coarse[-1])
```

File: tinyDA/chain.py [hunk 14]
```diff
@@ -489,6 +523,14 @@ def _update_error_model(self):
             )
         self.chain_coarse[-1] = self.posterior_coarse.update_link(self.chain_coarse[-1])
 
+    def _get_random_proposal_index(self):
+        random_proposal_index = np.random.randint(-self.subchain_length,0) #prob does not correspond to correct state
```
[the remark this answers is not in the document]
Comment (louisekluge) [reply tinyDA/chain.py:527]: Yes, I think so.

File: tinyDA/chain.py (hunk 7 above)
Comment (louisekluge) [reply tinyDA/chain.py]: Oooh, I see. Sorry, I got confused!

**Review by mikkelbue — CHANGES_REQUESTED** [review]
Looks great. 

One final thing, I think the `self.chain_coarse[-1].model_output` in the method `_get_state_dependent_acceptance` line 448 has to be `self.promoted_coarse[-1].model_output`, since that compute the state-dependent difference between the coarse and fine model, so that would have to be the model output for the same set of parameters.

**Comment by louisekluge** [comment]
Ok, fixed it. Thanks for keeping an eye on these kinds of errors, I've unfortunately lost track of them somewhere.

**Comment by mikkelbue** [comment]
@louisekluge Looks like it's working!

I did try to run parallel sampling with `ray`, and that didn't work. I don't suppose you have tried running this with `ray`? Basically, we need to add this in to the class `ParallelDAChain` in the file `ray.py`to get the `randomize_subchain_length` to be passed through to the underlying `DAChain.py` when running in parallel.

**Comment by louisekluge** [comment]
Aah, I see. No I didn't try with ray on my work laptop

**Comment by mikkelbue** [comment]
> Aah, I see. No I didn't try with ray on my work laptop

@louisekluge no worries, should be a small job. It's just some boilerplate to pass the argument through.

File: tinyDA/ray.py [hunk 15]
```diff
@@ -132,6 +136,7 @@ def __init__(
                 self.proposal[i],
                 self.subchain_length,
                 self.initial_parameters[i],
+                self.randomize_subchain_lenght,
```
Comment (mikkelbue) [inline tinyDA/ray.py:139]:
```suggestion
                self.randomize_subchain_length,
```

**Review by mikkelbue — CHANGES_REQUESTED** [review]
I found a spelling error but it should otherwise be fine.

File: tinyDA/ray.py (hunk 15 above)
Comment (louisekluge) [reply tinyDA/ray.py]: Thanks for pointing it out!

**Comment by louisekluge** [comment]
Ok, something is still not right, either within chain.py or sampler.py or ray.py? When I test the DA with AEM notebook, and sample with 2 chains in parallel and randomize_subchain_length, then I get a TypeError
Screenshot from 2024-10-07 12-04-31

**Comment by mikkelbue** [comment]
@louisekluge It's probably because you are passing in the `randomize_subchain_length` as the 5th parameter in `sampler.py`:
```
    chains = ParallelDAChain(
        posteriors[0],
        posteriors[1],
        proposal,
        subchain_length,
        randomize_subchain_length,
        n_chains,
        initial_parameters,
        adaptive_error_model,
        store_coarse_chain,
    )
```
but `ParallelDAChain` in `ray.py` expects it to be the 7th parameter:
```
    def __init__(
        self,
        posterior_coarse,
        posterior_fine,
        proposal,
        subchain_length=1,
        n_chains=2,
        initial_parameters=None,
        randomize_subchain_length=False,
        adaptive_error_model=None,
        store_coarse_chain=True,
    ):
```

The simplest would probably be to make the order of parameters for `ParallelDaChain` the same as for `DAChain`.

**Comment by louisekluge** [comment]
Indeed, that was it, notebooks run through now. Now it's time to look at the estimator.

Also, I modified the model in the DA with AEM notebook so that instead of 
      else:
          return np.nan, False
it now returns instead the same thing as for self.y.success, but multiplied by zero.
      else:
          return self.y.y.flatten()*0, False
Is that something you'd want to include in the main version?

**Comment by mikkelbue** [comment]
> it now returns instead the same thing as for self.y.success, but multiplied by zero.

I think we could do `return np.zeros_like(self.y.y.flatten()), False` instead and that would be a bit more idiomatic NumPy.

**Comment by mikkelbue** [comment]
Once that is fixed in the notebook let's get this merged and address the estimator in another PR.
